jdaviz/core/helpers.py

```diff
@@ -402,6 +407,87 @@ def show_in_new_tab(self, title=None):  # pragma: no cover
                       DeprecationWarning)
         return self.show(loc="sidecar:tab-after", title=title)
 
+    def get_data(self, data_label=None, cls=None, subset_to_apply=None, statistic=None):
+        """
+        Returns data with name equal to `data_label` of type `cls` with subsets applied from
+        `subsets_to_apply`.
+
+        Parameters
+        ----------
+        data_label : str, optional
+            Provide a label to retrieve a specific data set from data_collection.
+        cls : `~specutils.Spectrum1D`, 'default', None
+            The type that data will be returned as.
+        subset_to_apply : str, optional
+            Subset that is to be applied to data before it is returned.
+        statistic : {'minimum', 'maximum', 'mean', 'median', 'sum', 'percentile'}, optional
+            The statistic to use to collapse the dataset.
+
+        Returns
+        -------
+        data : `cls`
+            Data is returned as type `cls` with subsets applied if applicable.
+
+        """
+        if data_label and data_label not in self.app.data_collection.labels:
+            raise ValueError(f'{data_label} not in {self.app.data_collection.labels}.')
+        elif not data_label and len(self.app.data_collection) > 1:
+            raise ValueError('data_label must be set if more than'
+                             ' one data exists in data_collection.')
+        elif not data_label and len(self.app.data_collection) == 1:
+            data_label = self.app.data_collection[0].label
+
+        # TODO: Previously, this would use viewer.default_class. Do we want to pass
+        #  a viewer reference in order to enable this? Or move this to the base
+        #  viewer class?
+        cls = Spectrum1D if cls == 'default' else cls
+
+        if cls is not None and not isclass(cls):
+            raise TypeError(
+                "cls in get_data_from_viewer must be a class, None, or "
+                "the 'default' string.")
+        data = self.app.data_collection[data_label]
+
+        if cls is not None:
+            # If data is one-dimensional, assume that it can be
+            #  collapsed via the defined statistic
+            if 'Trace' in data.meta:
+                data = data.get_object()
+            elif cls == Spectrum1D:
+                data = data.get_object(cls=cls, statistic=statistic)
+            else:
+                data = data.get_object(cls=cls)
+        # If the shape of the data is 2d, then use CCDData as the
+        #  output data type
+        elif len(data.shape) == 2:
+            cls = CCDData
+            data = data.get_object(cls=cls)
+        elif len(data.shape) in [1, 3]:
+            cls = Spectrum1D
+            data = data.get_object(cls=cls, statistic=statistic)
+
+        if not subset_to_apply or not cls:
+            return data
```

Comment:
The last two `elif` statements are needed outside the main block to set the `cls` to a valid class if it is set to 'None` from the start. With that, `cls=None` will be the default (instead of `cls='default'`) and the code will figure out which to use based on the shape of the glue data object. This works with Imviz, Cubeviz, and Specviz.

Reply:
ok, I see, but `data` itself isn't actually used unless `subset_to_apply` is not provided, right?  Is there any performance concern for all the `get_object` calls that are effectively discarded?

Reply:
How does the newest implementation look to you?

Reply:
I like it - its a bit more verbose, but I think the logic paths are more clear this way.  Thanks!

Reply:
Ok, I'll work on adding some tests to increase coverage. Thanks for the review!